test/functional/p2p_getaddr_caching.py

```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env python3
+# Copyright (c) 2020 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test addr response caching"""
+
+from test_framework.messages import (
+    CAddress,
+    NODE_NETWORK,
+    NODE_WITNESS,
+    msg_addr,
+    msg_getaddr,
+)
+from test_framework.mininode import (
+    P2PInterface,
+    mininode_lock
+)
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import (
+    assert_equal,
+    wait_until
+)
+import time
+
+def gen_addrs(n):
+    addrs = []
+    for i in range(n):
+        addr = CAddress()
+        addr.time = int(time.time())
+        addr.nServices = NODE_NETWORK | NODE_WITNESS
+        # Use first octets to occupy different AddrMan buckets
+        first_octet = i >> 8
+        second_octet = i % 256
+        addr.ip = "{}.{}.1.1".format(first_octet, second_octet)
+        addr.port = 8333
+        addrs.append(addr)
+    return addrs
+
+class AddrReceiver(P2PInterface):
+
+    def __init__(self):
+        super().__init__()
+        self.received_addrs = None
+
+    def on_addr(self, message):
+        self.received_addrs = []
+        for addr in message.addrs:
+            self.received_addrs.append(addr.ip)
+
+    def received_addr(self):
+        return self.received_addrs is not None
+
+
+class AddrTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.setup_clean_chain = False
+        self.num_nodes = 1
+
+    def run_test(self):
+        self.log.info('Create connection that sends and requests addr messages')
+        addr_source = self.nodes[0].add_p2p_connection(P2PInterface())
+
+        msg_send_addrs = msg_addr()
+        self.log.info('Fill peer AddrMan with a lot of records')
+        addrs = gen_addrs(10000)
+        for i in range(10):
+            msg_send_addrs.addrs = addrs[i * 1000:(i + 1) * 1000]
+            addr_source.send_and_ping(msg_send_addrs)
```

Comment:
This doesn't appear to fill the addrman with 10,000 address records. If I run the `getnodeaddresses` RPC after this repeatedly and keep a set of all of the address records returned, I only get to around 3500 (it varies on each test run).

Reply:
Sure, but my goal here was to make it more than 1,000; just in case. Any particular suggestion?

Reply:
Because of the limit of returning 23% of records in GetAddr(), there need to be ~4350 records for the cache to be filled with 1000 records.

I haven't looked into how CAddress records in ADDR messages are saved to AddrMan, so I can't offer any insight into why only ~3500 are being saved here.

Reply:
Sorry, I'm still not sure what we're trying to accomplish here... I think as long as we use 1,000 and not 2,500 for cache size, either cache size (less or equal then 1,000) is fine.

I just chose this 10,000 value so that it's *big enough*, just to hit some potential corner case.  I don't see a real reason to go beyond that (for example, to have a cache size of 1,000). I can probably achieve that, but right now I just don't see the motivation.

Reply:
The motivation is that the tests should test the limits of the new functionality. The cache can have up to 1000 entries in it. What's the behavior like when that cache is full?

The description of this PR says:

> cache responses to GETADDR on a daily basis, so that an attacker gets at most 1,000 records per day

The test should test that an attacker is actually only able to get 1,000 records per day.

Reply:
I updated the test and I think these 2 checks are sort of sufficient here:
1. All responses during the day are same
2. They are less than 1000 records
I think (1) is much more interesting then (2), and the second part of the test confirms that (1) is only day-long cache.

I'll take a look why it's not exact cache size of 1000 later, but I still think it's not that substantial.

Reply:
Alright, so the issue is that we map every *source* of the ADDRs to 64 buckets 64 slots each -> one *source* can add at most 64*64=4096 records. This results in at most 4096*0.23=942 records returned from AddrMan.

As far as I know, there is currently no good way to work around this per-source limit while testing, so leaving it as is for now.
As I mention above, I think this is not critical for this test.

Reply:
ok, sounds reasonable. It's a shame that our testing framework won't allow us to fully test the edges of this, but that shouldn't hold this PR up.